flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkImageBuilder.java

```diff
@@ -0,0 +1,240 @@
+
+    private final Map<Path, Path> filesToCopy = new HashMap<>();
+
+    private String imageName = DEFAULT_IMAGE_NAME;
+    private String imageNameSuffix;
+    private Path flinkDist = FileUtils.findFlinkDist();
+    private String javaVersion;
+    private Configuration conf;
+    private Duration timeout = DEFAULT_TIMEOUT;
+    private String startupCommand;
+
+    /**
+     * Sets the name of building image.
+     *
+     * <p>If the name is not specified, {@link #DEFAULT_IMAGE_NAME} will be used.
+     */
+    public FlinkImageBuilder setImageName(String imageName) {
+        this.imageName = imageName;
+        return this;
+    }
+
+    /**
+     * Sets the path of flink-dist directory.
+     *
+     * <p>If path is not specified, the dist directory under current project will be used.
+     */
+    public FlinkImageBuilder setFlinkDistPath(Path flinkDist) {
+        this.flinkDist = flinkDist;
+        return this;
+    }
+
+    /**
+     * Sets JDK version in the image.
+     *
+     * <p>If version is not specified, the JDK version of the current JVM will be used.
+     */
+    public FlinkImageBuilder setJavaVersion(String javaVersion) {
+        this.javaVersion = javaVersion;
+        return this;
+    }
+
+    /**
+     * Sets Flink configuration. This configuration will be used for generating flink-conf.yaml for
+     * configuring JobManager and TaskManager.
+     */
+    public FlinkImageBuilder setConfiguration(Configuration conf) {
+        this.conf = conf;
+        return this;
+    }
+
+    /** Copies file into the image. */
+    public FlinkImageBuilder copyFile(Path localPath, Path containerPath) {
+        filesToCopy.put(localPath, containerPath);
+        return this;
+    }
+
+    /** Sets timeout for building the image. */
+    public FlinkImageBuilder setTimeout(Duration timeout) {
+        this.timeout = timeout;
+        return this;
+    }
+
+    /** Use this image for building a JobManager. */
+    public FlinkImageBuilder asJobManager() {
+        checkStartupCommandNotSet();
+        this.startupCommand = "flink/bin/jobmanager.sh start-foreground && tail -f /dev/null";
+        this.imageNameSuffix = "jobmanager";
+        return this;
+    }
+
+    /** Use this image for building a TaskManager. */
+    public FlinkImageBuilder asTaskManager() {
+        checkStartupCommandNotSet();
+        this.startupCommand = "flink/bin/taskmanager.sh start-foreground && tail -f /dev/null";
+        this.imageNameSuffix = "taskmanager";
+        return this;
+    }
+
+    /** Use a custom command for starting up the container. */
+    public FlinkImageBuilder useCustomStartupCommand(String command) {
+        checkStartupCommandNotSet();
+        this.startupCommand = command;
+        this.imageNameSuffix = "custom";
+        return this;
+    }
+
+    /** Build the image. */
+    public ImageFromDockerfile build() throws TimeoutException, IOException {
```

Comment:
WDYT about simplifying the Exception handling by only throwing an `ImageBuildException` during runtime from this method?

The exception can already contain if it is the taskmanager or jobmanager and none of the users needs to handle and wrap it again.